libsignal-protocol/src/crypto/openssl.rs

```diff
@@ -38,20 +39,36 @@ impl OpenSSLCrypto {
             (SignalCipherType::AesCbcPkcs5, 32) => Cipher::aes_256_cbc(),
             _ => unreachable!(),
         };
-
         let block_size = signal_cipher_type.block_size();
-        let mut result = Vec::with_capacity(data.len() + block_size);
+        // turns out that we have to fill the buffer with some initial value
+        // to pass it to the openssl library.
+        // also it depends on the mode, in the AesCtr (aka stream cipher) mode
+        // we had to provide the same exact size as the input.
+        // in the AesCbs, the buffer should be `data.len() + blocks_ize`
+        // but there is a small problem here, that the returned value **the result buffer**
+        // has a `data.len()` 0's tail at the end of the buffer !
+        // for example, if the `data = [1, 2, 3, 4]`
+        // the native (aka DefaultCrypto) will result for example
+        // [70, 108, 98, 83, 33, 54, 241, 25, 86, 110, 44, 34, 228, 183, 215, 251]
+        // and the openssl (aka OpenSSLCrypto) will result
+        // [70, 108, 98, 83, 33, 54, 241, 25, 86, 110, 44, 34, 228, 183, 215, 251, 0, 0, 0, 0]
+        // note the [..., 0, 0, 0, 0] at the end, it always has the same len of the input `data`.
+        // see `test_crypter` unit test in the [`./crypto/mod.rs`]
+        //
+        // FIXME (@shekohex): Find why openssl has that behavior
+        let mut result = match cipher {
+            SignalCipherType::AesCtrNoPadding => vec![0u8; data.len()],
+            SignalCipherType::AesCbcPkcs5 => vec![0u8; data.len() + block_size],
+        };
         let mut crypter = Crypter::new(signal_cipher_type, mode, key, Some(iv))
             .map_err(|_e| InternalError::Unknown)?;
-
         crypter
             .update(data, &mut result)
```

Comment:
Reading throw source code of `crypter.update(...)` i found this
https://github.com/sfackler/rust-openssl/blob/ef86438a109f7f7683bde078563750a3ed02994d/openssl/src/symm.rs#L487-L505

especially this comment
```rust
    /// # Panics
    ///
    /// Panics for stream ciphers if `output.len() < input.len()`.
    ///
    /// Panics for block ciphers if `output.len() < input.len() + block_size`,
    /// where `block_size` is the block size of the cipher (see `Cipher::block_size`).
```

> stream ciphers (aka CTR), block ciphers (aka CBC), FYI, CTR has a block_size of 1, so no problem at all.

Reply:
Do you think this is something we'll need to worry about? We could always document it and call it a day, otherwise there's probably a way to ensure the assumption is valid at a higher level...